Spurious 'conflicting implementations' error when specializing a type with a compilation error

The following code:

```rust
#![feature(specialization)]

struct BadStruct {
    err: MissingType
}

trait MyTrait<T> {
    fn foo();
}

impl<T, D> MyTrait<T> for D {
    default fn foo() {}
}

impl<T> MyTrait<T> for BadStruct {
    fn foo() {}
}
```

gives the following errors:

```rust
error[E0412]: cannot find type `MissingType` in this scope
 --> src/lib.rs:4:10
  |
4 |     err: MissingType
  |          ^^^^^^^^^^^ not found in this scope

error[E0119]: conflicting implementations of trait `MyTrait<_>` for type `BadStruct`:
  --> src/lib.rs:15:1
   |
11 | impl<T, D> MyTrait<T> for D {
   | --------------------------- first implementation here
...
15 | impl<T> MyTrait<T> for BadStruct {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `BadStruct`

error: aborting due to 2 previous errors
```

For some reason, the fact that the definition of `BadStruct` has an error (`MissingType` is not defining) causes a 'conflicting implementations' error to be emitted when `BadStruct` has a specialized impl. If `MissingType` is changed to a type which actually exists (e.g. `()`), the 'conflicting implementations' error disappears.

I found this when working on `rustc` - a missing `use` statement caused 30 spurious specialization-related errors to be emiited, which all disappeared when I added the missing import.

Current output:
```
error[E0412]: cannot find type `MissingType` in this scope
 --> src/main.rs:5:10
  |
5 |     err: MissingType
  |          ^^^^^^^^^^^ not found in this scope

error: aborting due to previous error
```

The E0119 is no longer shown, marking as E-needs-test.

--- a/src/test/ui/specialization/issue-68830-spurious-diagnostics.rs
+++ b/src/test/ui/specialization/issue-68830-spurious-diagnostics.rs
@@ -1,0 +1,23 @@
+// A regression test for #68830. This checks we don't emit
+// a verbose `conflicting implementations` error.
+
+#![feature(specialization)]
+#![allow(incomplete_features)]
+
+struct BadStruct {
+    err: MissingType //~ ERROR: cannot find type `MissingType` in this scope
+}
+
+trait MyTrait<T> {
+    fn foo();
+}
+
+impl<T, D> MyTrait<T> for D {
+    default fn foo() {}
+}
+
+impl<T> MyTrait<T> for BadStruct {
+    fn foo() {}
+}
+
+fn main() {}

--- a/src/test/ui/specialization/issue-68830-spurious-diagnostics.stderr
+++ b/src/test/ui/specialization/issue-68830-spurious-diagnostics.stderr
@@ -1,0 +1,9 @@
+error[E0412]: cannot find type `MissingType` in this scope
+  --> $DIR/issue-68830-spurious-diagnostics.rs:8:10
+   |
+LL |     err: MissingType
+   |          ^^^^^^^^^^^ not found in this scope
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0412`.

